framework/src/main/java/org/checkerframework/framework/type/treeannotator/PropagationTreeAnnotator.java

```diff
@@ -237,7 +237,15 @@ public Void visitBinary(BinaryTree node, AnnotatedTypeMirror type) {
                 qualHierarchy.leastUpperBounds(
                         argTypes.first.getEffectiveAnnotations(),
                         argTypes.second.getEffectiveAnnotations());
-        type.addMissingAnnotations(lubs);
+
+        if (TreeUtils.isBinaryComparison(node)) {
+            // When we have binary comparison, the result type (boolean) can be different
+            // from the operands' types. So we need to apply the bounds of boolean to the
+            // lubs.
+            addAnnoOrBound(type, lubs);
```

Comment:
Is this helper method still useful?
The code might be easier if you directly re-assign lubs here and then use `type.addMissingAnnotations(lubs)` in both branches (i.e. move out of the else).
There are two other calls, so maybe it's still useful.

Reply:
Removed this `addAnnoOrBound` invocation and re-assigned `lubs`. Thanks!